runtime/interpreter/value_link.go

```diff
@@ -0,0 +1,311 @@
+/*
+ * Cadence - The resource-oriented smart contract programming language
+ *
+ * Copyright Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package interpreter
+
+import (
+	"github.com/onflow/atree"
+
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/cadence/runtime/errors"
+)
+
+// TODO: remove once migrated
+
+// Deprecated: LinkValue
+type LinkValue interface {
+	Value
+	isLinkValue()
+}
+
+// Deprecated: PathLinkValue
+type PathLinkValue struct {
+	Type       StaticType
+	TargetPath PathValue
+}
+
+var EmptyPathLinkValue = PathLinkValue{}
+
+var _ Value = PathLinkValue{}
+var _ atree.Value = PathLinkValue{}
+var _ EquatableValue = PathLinkValue{}
+var _ LinkValue = PathLinkValue{}
+
+func (PathLinkValue) isValue() {}
+
+func (PathLinkValue) isLinkValue() {}
+
+func (v PathLinkValue) Accept(_ *Interpreter, _ Visitor) {
+	panic(errors.NewUnreachableError())
+}
+
+func (v PathLinkValue) Walk(_ *Interpreter, _ func(Value)) {
+	panic(errors.NewUnreachableError())
+}
+
+func (v PathLinkValue) StaticType(_ *Interpreter) StaticType {
+	panic(errors.NewUnreachableError())
```

Comment:
To be clear, we don't need implementations for these because they are getting removed once the migration is over?

Reply:
Correct. The old values are added back to be able to decode them from the state that gets migrated (e.g decoding function), and to use them during the migration (e.g. equality function). Just the code that is needed was added, all other functions got stubbed out to panic if they are ever used (which should never happen during migration).